### PR TITLE
Rp2040 smp

### PR DESCRIPTION
**Pull request by kilograham**
Adds support for RP2040 (with SMP) using the Raspberry Pi Pico SDK. FreeRTOS can be run on either single core, or in both cores in SMP mode. SDK semaphores, queues, mutexes, and sleep functions can be used freely to/from FreeRTOS tasks

This PR contains two commits; one which is identical to #341 (i.e. the changes required for RP2040 support without SMP), and a second commit that adds SMP support. It is probably advisable therefore to review #341 first.

### CODE REVIEW

**Comment by kilograham** [comment]
i see that the RP2040 non SMP PR has been merged (thank you) I will rebase this on top of that

File: include/stack_macros.h [hunk 1]
```diff
@@ -43,13 +43,21 @@
...
+#endif
+
 #if ( ( configCHECK_FOR_STACK_OVERFLOW == 1 ) && ( portSTACK_GROWTH < 0 ) )
 
 /* Only the current stack state is to be checked. */
     #define taskCHECK_FOR_STACK_OVERFLOW()                                                            \
     {                                                                                                 \
         /* Is the currently saved stack pointer within the stack limit? */                            \
-        if( pxCurrentTCB->pxTopOfStack <= pxCurrentTCB->pxStack )                                     \
+        if( pxCurrentTCB->pxTopOfStack <= pxCurrentTCB->pxStack + portSTACK_LIMIT_PADDING)            \
```
Comment (aggarg) [inline include/stack_macros.h:60]:
```suggestion
        if( pxCurrentTCB->pxTopOfStack <= pxCurrentTCB->pxStack + portSTACK_LIMIT_PADDING )           \
```

File: include/stack_macros.h [hunk 2]
```diff
@@ -65,7 +73,7 @@
     {                                                                                                 \
                                                                                                       \
         /* Is the currently saved stack pointer within the stack limit? */                            \
-        if( pxCurrentTCB->pxTopOfStack >= pxCurrentTCB->pxEndOfStack )                                \
+        if( pxCurrentTCB->pxTopOfStack >= pxCurrentTCB->pxEndOfStack - portSTACK_LIMIT_PADDING)       \
```
Comment (aggarg) [inline include/stack_macros.h:76]:
```suggestion
        if( pxCurrentTCB->pxTopOfStack >= pxCurrentTCB->pxEndOfStack - portSTACK_LIMIT_PADDING )      \
```

**Review by aggarg — APPROVED** [review]

**Pull request merged**